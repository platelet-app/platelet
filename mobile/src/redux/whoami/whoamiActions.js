export const GET_WHOAMI_REQUEST = "GET_WHOAMI_REQUEST";
export const REFRESH_WHOAMI_REQUEST = "REFRESH_WHOAMI_REQUEST";
export const GET_WHOAMI_SUCCESS = "GET_WHOAMI_SUCCESS";
export const GET_WHOAMI_FAILURE = "GET_WHOAMI_FAILURE";
export const CLEAR_WHOAMI = "CLEAR_WHOAMI";
export const INIT_WHOAMI_OBSERVER = "INIT_WHOAMI_OBSERVER";
export const SET_TENANT_ID = "SET_TENANT_ID";

export function getWhoamiRequest() {
    return { type: GET_WHOAMI_REQUEST };
}

export function refreshWhoamiRequest() {
    return { type: REFRESH_WHOAMI_REQUEST };
}

export function clearWhoami() {
    return { type: CLEAR_WHOAMI };
}

export function getWhoamiSuccess(data) {
    return { type: GET_WHOAMI_SUCCESS, data };
}

export function initWhoamiObserver(whoamiId) {
    return { type: INIT_WHOAMI_OBSERVER, whoamiId };
}

export function getWhoamiFailure(error) {
    return { type: GET_WHOAMI_FAILURE, error };
}

export function setTenantId(data) {
    return { type: SET_TENANT_ID, data };
}
