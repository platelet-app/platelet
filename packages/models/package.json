{
    "name": "@platelet-app/models",
    "version": "1.0.1",
    "description": "DataStore models for the platelet.app project",
    "homepage": "https://platelet.app",
    "bugs": {
        "url": "https://github.com/platelet-app/platelet/issues"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/platelet-app/platelet.git"
    },
    "license": "Apache-2.0",
    "author": "Theo Cranmore <theo@platelet.app>",
    "type": "module",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "build": "cd ../.. && amplify api gql-compile && amplify codegen models && cd packages/models && cp ../../models/* dist",
        "prepublishOnly": "npm version patch"
    }
}
