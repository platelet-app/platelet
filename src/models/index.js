// @ts-check
import { initSchema } from '@aws-amplify/datastore';
import { schema } from './schema';

const Role = {
  "USER": "USER",
  "COORDINATOR": "COORDINATOR",
  "RIDER": "RIDER",
  "ADMIN": "ADMIN"
};

const CommentVisibility = {
  "EVERYONE": "EVERYONE",
  "ME": "ME"
};

const TaskStatus = {
  "NEW": "NEW",
  "ACTIVE": "ACTIVE",
  "PICKED_UP": "PICKED_UP",
  "DROPPED_OFF": "DROPPED_OFF",
  "CANCELLED": "CANCELLED",
  "REJECTED": "REJECTED",
  "ABANDONED": "ABANDONED",
  "COMPLETED": "COMPLETED"
};

const Priority = {
  "HIGH": "HIGH",
  "MEDIUM": "MEDIUM",
  "LOW": "LOW"
};

const DeliverableTypeIcon = {
  "BUG": "BUG",
  "CHILD": "CHILD",
  "DOCUMENT": "DOCUMENT",
  "EQUIPMENT": "EQUIPMENT",
  "OTHER": "OTHER"
};

const DeliverableUnit = {
  "NONE": "NONE",
  "LITER": "LITER",
  "MILLILITER": "MILLILITER",
  "GRAM": "GRAM",
  "ITEM": "ITEM",
  "BOX": "BOX"
};

const { User, PossibleRiderResponsibilities, RiderResponsibility, Comment, TaskAssignee, Task, Location, Handover, Deliverable, DeliverableType, VehicleAssignment, Vehicle, Tenant, Statistics, SendFeedback, AddressAndContactDetails, S3Object } = initSchema(schema);

export {
  User,
  PossibleRiderResponsibilities,
  RiderResponsibility,
  Comment,
  TaskAssignee,
  Task,
  Location,
  Handover,
  Deliverable,
  DeliverableType,
  VehicleAssignment,
  Vehicle,
  Tenant,
  Role,
  CommentVisibility,
  TaskStatus,
  Priority,
  DeliverableTypeIcon,
  DeliverableUnit,
  Statistics,
  SendFeedback,
  AddressAndContactDetails,
  S3Object
};