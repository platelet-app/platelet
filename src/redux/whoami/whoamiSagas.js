import { call, put, take, takeLatest } from "redux-saga/effects";
import {
    GET_WHOAMI_REQUEST,
    getWhoamiFailure,
    getWhoamiSuccess,
    REFRESH_WHOAMI_REQUEST,
    setTenantId,
    INIT_WHOAMI_OBSERVER,
    initWhoamiObserver,
} from "./whoamiActions";
import API from "@aws-amplify/api";
import { Auth, DataStore, syncExpression } from "aws-amplify";
import * as models from "../../models";
import * as queries from "../../graphql/queries";
import { NotFound } from "http-errors";
import { userRoles } from "../../apiConsts";
import { eventChannel } from "redux-saga";
import dataStoreConflictHandler from "./dataStoreConflictHandler";

const fakeUser = {
    id: "offline",
    name: "offline",
    username: "offline",
    displayName: "Offline User",
    roles: Object.values(userRoles),
    dateOfBirth: null,
    profilePictureURL: null,
    profilePictureThumbnailURL: null,
};

const testUserModel = new models.User({
    name: "whoami",
    displayName: "Mock User",
    roles: Object.values(userRoles),
    dateOfBirth: null,
    profilePictureURL: null,
    profilePictureThumbnailURL: null,
});
const testUser = { ...testUserModel, id: "whoami" };

function listener(userId) {
    return eventChannel((emitter) => {
        let observer = { unsubscribe: () => {} };
        observer = DataStore.observe(models.User, userId).subscribe(
            ({ element }) => emitter(element)
        );
        return () => {
            observer.unsubscribe();
        };
    });
}

function* whoamiObserver(action) {
    const channel = yield call(listener, action.whoamiId);
    try {
        while (true) {
            const result = yield take(channel);
            yield put(getWhoamiSuccess(result));
        }
    } finally {
        console.log("stopping whoami observer");
        channel.close();
    }
}

export function* watchInitWhoamiObserver() {
    yield takeLatest(INIT_WHOAMI_OBSERVER, whoamiObserver);
}

function* getWhoami() {
    if (process.env.NODE_ENV === "test") {
        yield put(getWhoamiSuccess(testUser));
        return;
    }
    if (
        process.env.REACT_APP_DEMO_MODE === "true" ||
        process.env.REACT_APP_OFFLINE_ONLY === "true"
    ) {
        const existingUser = yield call(
            [DataStore, DataStore.query],
            models.User,
            (u) => u.name("eq", "offline")
        );
        if (existingUser.length === 0) {
            let userResult = fakeUser;
            const userModel = yield new models.User(userResult);
            const newFakeUser = yield call(
                [DataStore, DataStore.save],
                userModel
            );
            yield put(getWhoamiSuccess(newFakeUser));
            yield put(initWhoamiObserver(newFakeUser.id));
        } else {
            yield put(getWhoamiSuccess(existingUser[0]));
            yield put(initWhoamiObserver(existingUser[0].id));
        }
        yield put(setTenantId("offline"));
    } else {
        try {
            const loggedInUser = yield call([
                Auth,
                Auth.currentAuthenticatedUser,
            ]);
            let result;
            if (loggedInUser) {
                if (
                    !loggedInUser.attributes ||
                    !loggedInUser.attributes["custom:tenantId"]
                ) {
                    yield put(getWhoamiFailure("No tenantId"));
                    return;
                }

                const tenantId = loggedInUser.attributes["custom:tenantId"];
                const modelsToSync = [];
                for (const model of Object.values(models)) {
                    if (
                        [
                            "Task",
                            "User",
                            "TaskAssignee",
                            "RiderResponsibility",
                            "Comment",
                            "Location",
                            "Vehicle",
                            "Deliverable",
                            "DeliverableType",
                            "PossibleRiderResponsibilities",
                        ].includes(model.name)
                    ) {
                        modelsToSync.push(model);
                    }
                }

                yield call([DataStore, DataStore.configure], {
                    errorHandler: (err) => {
                        console.log("DataStore error:", err);
                        console.log("Cause:", err.cause);
                    },
                    syncExpressions: [
                        ...modelsToSync.map((model) =>
                            syncExpression(model, (m) =>
                                m.tenantId("eq", tenantId)
                            )
                        ),
                        syncExpression(models.Tenant, (m) =>
                            m.id("eq", tenantId)
                        ),
                    ],
                    conflictHandler: dataStoreConflictHandler,
                });
                result = yield call(
                    [DataStore, DataStore.query],
                    models.User,
                    (t) => t.cognitoId("eq", loggedInUser.attributes.sub)
                );
                if (result && result.length === 0) {
                    result = yield call([API, API.graphql], {
                        query: queries.getUserByCognitoId,
                        variables: { cognitoId: loggedInUser.attributes.sub },
                    });
                    if (
                        result &&
                        result.data &&
                        result.data.getUserByCognitoId &&
                        result.data.getUserByCognitoId.items &&
                        result.data.getUserByCognitoId.items.length > 0
                    ) {
                        yield put(
                            getWhoamiSuccess(
                                result.data.getUserByCognitoId.items[0]
                            )
                        );
                        yield put(
                            initWhoamiObserver(
                                result.data.getUserByCognitoId.items[0].id
                            )
                        );
                    } else {
                        throw new NotFound("Could not find logged in user");
                    }
                } else {
                    yield put(getWhoamiSuccess(result[0]));
                    yield put(initWhoamiObserver(result[0].id));
                }
                yield put(setTenantId(tenantId));
            } else {
                yield put(
                    getWhoamiFailure(
                        new NotFound("Could not find logged in user")
                    )
                );
            }
        } catch (error) {
            console.log(error);
            yield put(getWhoamiFailure(error));
        }
    }
}

export function* watchGetWhoami() {
    yield takeLatest(GET_WHOAMI_REQUEST, getWhoami);
}

export function* watchRefreshWhoami() {
    yield takeLatest(REFRESH_WHOAMI_REQUEST, getWhoami);
}
