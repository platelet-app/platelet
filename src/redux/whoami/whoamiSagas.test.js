import { API, Auth, DataStore } from "aws-amplify";
import { testFunctions } from "./whoamiSagas";
import { runSaga } from "redux-saga";
import * as awsAmplify from "aws-amplify";
import * as models from "../../models";
import * as whoamiActions from "./whoamiActions";

const fakeUser = {
    cognitoId: "someCognitoId",
    tenantId: "someTenantId",
    id: "someUserId",
};

const fakeCognitoResponse = {
    attributes: {
        sub: "someCognitoId",
    },
};

jest.mock("../../aws-exports", () => ({
    default: {
        aws_appsync_graphqlEndpoint: "https://another.com/graphql",
    },
}));

jest.mock("uuid", () => ({ v4: () => "someUUID" }));

describe("whoamiSagas", () => {
    beforeEach(() => {
        process.env.REACT_APP_DEMO_MODE = "false";
        process.env.REACT_APP_OFFLINE_ONLY = "false";
    });
    afterEach(async () => {
        jest.restoreAllMocks();
        await DataStore.clear();
    });
    it("get the user data from the API and save the tenantId", async () => {
        const dispatched = [];
        jest.spyOn(Auth, "currentAuthenticatedUser").mockResolvedValue(
            fakeCognitoResponse
        );
        jest.spyOn(Auth, "currentSession").mockResolvedValue({
            getIdToken: () => ({ getJwtToken: () => "someToken" }),
        });
        const apiSpy = jest.spyOn(API, "graphql").mockResolvedValue({
            data: { getUserByCognitoId: { items: [fakeUser] } },
        });
        jest.spyOn(Storage.prototype, "getItem")
            .mockReturnValueOnce(null)
            .mockReturnValueOnce(
                '{"aws_appsync_graphqlEndpoint": "https://amazonaws.com/graphq"}'
            );
        const setSpy = jest.spyOn(Storage.prototype, "setItem");
        const syncExpressionSpy = jest.spyOn(awsAmplify, "syncExpression");
        const mockFetchPromise = Promise.resolve({
            json: () =>
                Promise.resolve({
                    data: {
                        getUserByCognitoId: {
                            items: [{ tenantId: "someTenantId" }],
                        },
                    },
                }),
        });
        const fetchSpy = jest
            .spyOn(global, "fetch")
            .mockImplementation(() => mockFetchPromise);

        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();

        expect(dispatched).toMatchSnapshot();
        expect(setSpy).toHaveBeenCalledWith("userTenantId", "someTenantId");
        for (const m of Object.values(models)) {
            if (!m.hasOwnProperty("copyOf")) continue;
            expect(syncExpressionSpy).toHaveBeenCalledWith(
                m,
                expect.any(Function)
            );
        }
        expect(fetchSpy.mock.calls).toMatchSnapshot();
        expect(apiSpy.mock.calls).toMatchSnapshot();
    });
    it("get the user data from the API and save the tenantId but get config from aws-exports", async () => {
        process.env.REACT_APP_TENANT_GRAPHQL_ENDPOINT = undefined;
        const dispatched = [];
        jest.spyOn(Auth, "currentAuthenticatedUser").mockResolvedValue(
            fakeCognitoResponse
        );
        jest.spyOn(Auth, "currentSession").mockResolvedValue({
            getIdToken: () => ({ getJwtToken: () => "someToken" }),
        });
        const apiSpy = jest.spyOn(API, "graphql").mockResolvedValue({
            data: { getUserByCognitoId: { items: [fakeUser] } },
        });
        jest.spyOn(Storage.prototype, "getItem")
            .mockReturnValueOnce(null)
            .mockReturnValueOnce(
                '{"aws_appsync_graphqlEndpoint": "https://amazonaws.com/graphq"}'
            );
        const setSpy = jest.spyOn(Storage.prototype, "setItem");
        const syncExpressionSpy = jest.spyOn(awsAmplify, "syncExpression");
        const mockFetchPromise = Promise.resolve({
            json: () =>
                Promise.resolve({
                    data: {
                        getUserByCognitoId: {
                            items: [{ tenantId: "someTenantId" }],
                        },
                    },
                }),
        });
        const fetchSpy = jest
            .spyOn(global, "fetch")
            .mockImplementation(() => mockFetchPromise);

        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();

        expect(dispatched).toMatchSnapshot();
        expect(setSpy).toHaveBeenCalledWith("userTenantId", "someTenantId");
        for (const m of Object.values(models)) {
            if (!m.hasOwnProperty("copyOf")) continue;
            expect(syncExpressionSpy).toHaveBeenCalledWith(
                m,
                expect.any(Function)
            );
        }
        expect(fetchSpy.mock.calls).toMatchSnapshot();
        expect(apiSpy.mock.calls).toMatchSnapshot();
    });
    it("get the user data from DataStore and the tenantId from localstorage", async () => {
        const dispatched = [];
        const mockWhoami = await DataStore.save(
            new models.User({
                displayName: "test",
                tenantId: "someTenant",
                cognitoId: "someCognitoId",
            })
        );
        jest.spyOn(Auth, "currentAuthenticatedUser").mockResolvedValue(
            fakeCognitoResponse
        );
        jest.spyOn(API, "graphql").mockResolvedValue({
            data: { getUserByCognitoId: { items: [fakeUser] } },
        });
        jest.spyOn(Storage.prototype, "getItem").mockReturnValue(
            "someTenantId"
        );

        const apiSpy = jest.spyOn(API, "graphql").mockResolvedValue({});
        const setSpy = jest.spyOn(Storage.prototype, "setItem");
        const syncExpressionSpy = jest.spyOn(awsAmplify, "syncExpression");

        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();

        const successAction = dispatched.find(
            (a) => a.type === whoamiActions.GET_WHOAMI_SUCCESS
        );
        expect(successAction.data).toEqual(mockWhoami);
        const initWhoamiObserverAction = dispatched.find(
            (a) => a.type === whoamiActions.INIT_WHOAMI_OBSERVER
        );
        expect(initWhoamiObserverAction).toEqual({
            type: whoamiActions.INIT_WHOAMI_OBSERVER,
            whoamiId: mockWhoami.id,
        });
        const filteredDispatched = dispatched.filter(
            (a) =>
                a.type !== whoamiActions.GET_WHOAMI_SUCCESS &&
                a.type !== whoamiActions.INIT_WHOAMI_OBSERVER
        );
        expect(filteredDispatched).toMatchSnapshot();
        expect(apiSpy).not.toHaveBeenCalled();
        expect(setSpy).not.toHaveBeenCalled();
        for (const m of Object.values(models)) {
            if (!m.hasOwnProperty("copyOf")) continue;
            expect(syncExpressionSpy).toHaveBeenCalledWith(
                m,
                expect.any(Function)
            );
        }
    });
    it("demo mode", async () => {
        process.env.REACT_APP_DEMO_MODE = "true";
        const dispatched = [];
        const mockWhoami = await DataStore.save(
            new models.User({
                displayName: "test",
                name: "offline",
                tenantId: "someTenant",
                cognitoId: "someCognitoId",
            })
        );
        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();
        const successAction = dispatched.find(
            (a) => a.type === whoamiActions.GET_WHOAMI_SUCCESS
        );
        expect(successAction.data).toEqual(mockWhoami);
        const initWhoamiObserverAction = dispatched.find(
            (a) => a.type === whoamiActions.INIT_WHOAMI_OBSERVER
        );
        expect(initWhoamiObserverAction).toEqual({
            type: whoamiActions.INIT_WHOAMI_OBSERVER,
            whoamiId: mockWhoami.id,
        });
        const filteredDispatched = dispatched.filter(
            (a) =>
                a.type !== whoamiActions.GET_WHOAMI_SUCCESS &&
                a.type !== whoamiActions.INIT_WHOAMI_OBSERVER
        );
        expect(filteredDispatched).toMatchSnapshot();
    });
    it("API failure", async () => {
        const dispatched = [];
        jest.spyOn(Auth, "currentAuthenticatedUser").mockResolvedValue(
            fakeCognitoResponse
        );
        jest.spyOn(Auth, "currentSession").mockResolvedValue({
            getIdToken: () => ({ getJwtToken: () => "someToken" }),
        });
        jest.spyOn(global, "fetch").mockRejectedValue(new Error("someError"));
        jest.spyOn(Storage.prototype, "getItem")
            .mockReturnValueOnce(null)
            .mockReturnValueOnce(
                '{"aws_appsync_graphqlEndpoint": "https://amazonaws.com/graphq"}'
            );
        const setSpy = jest.spyOn(Storage.prototype, "setItem");

        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();

        expect(dispatched).toMatchSnapshot();
        expect(setSpy).not.toHaveBeenCalled();
    });
    it("errors from API", async () => {
        const dispatched = [];
        jest.spyOn(Auth, "currentAuthenticatedUser").mockResolvedValue(
            fakeCognitoResponse
        );
        jest.spyOn(Auth, "currentSession").mockResolvedValue({
            getIdToken: () => ({ getJwtToken: () => "someToken" }),
        });

        const mockFetchPromise = Promise.resolve({
            json: () =>
                Promise.resolve({
                    errors: [{ message: "some error from the server" }],
                    data: {
                        getUserByCognitoId: {},
                    },
                }),
        });
        jest.spyOn(global, "fetch").mockImplementation(() => mockFetchPromise);
        jest.spyOn(Storage.prototype, "getItem")
            .mockReturnValueOnce(null)
            .mockReturnValueOnce(
                '{"aws_appsync_graphqlEndpoint": "https://amazonaws.com/graphq"}'
            );
        const setSpy = jest.spyOn(Storage.prototype, "setItem");

        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();

        expect(dispatched).toMatchSnapshot();
        expect(setSpy).not.toHaveBeenCalled();
    });
    it("can't find the user", async () => {
        const dispatched = [];
        jest.spyOn(Auth, "currentAuthenticatedUser").mockResolvedValue(
            fakeCognitoResponse
        );
        jest.spyOn(Auth, "currentSession").mockResolvedValue({
            getIdToken: () => ({ getJwtToken: () => "someToken" }),
        });
        jest.spyOn(Storage.prototype, "getItem")
            .mockReturnValueOnce(null)
            .mockReturnValueOnce(
                '{"aws_appsync_graphqlEndpoint": "https://amazonaws.com/graphq"}'
            );
        const setSpy = jest.spyOn(Storage.prototype, "setItem");
        const mockFetchPromise = Promise.resolve({
            json: () =>
                Promise.resolve({
                    data: {
                        getUserByCognitoId: {
                            items: [],
                        },
                    },
                }),
        });
        jest.spyOn(global, "fetch").mockImplementation(() => mockFetchPromise);

        await runSaga(
            {
                dispatch: (action) => dispatched.push(action),
                getState: () => ({ state: "test" }),
            },
            testFunctions.getWhoami
        ).toPromise();

        expect(dispatched).toMatchSnapshot();
        expect(setSpy).not.toHaveBeenCalled();
    });
});
