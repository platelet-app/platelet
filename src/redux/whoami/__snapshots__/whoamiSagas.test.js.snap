// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whoamiSagas API failure 1`] = `
Array [
  Object {
    "error": [Error: someError],
    "type": "GET_WHOAMI_FAILURE",
  },
]
`;

exports[`whoamiSagas can't find the user 1`] = `
Array [
  Object {
    "error": [NotFoundError: Could not find tenant id for user],
    "type": "GET_WHOAMI_FAILURE",
  },
]
`;

exports[`whoamiSagas demo mode 1`] = `
Array [
  Object {
    "data": "offline",
    "type": "SET_TENANT_ID",
  },
]
`;

exports[`whoamiSagas get the user data from DataStore and the tenantId from localstorage 1`] = `
Array [
  Object {
    "data": "someTenantId",
    "type": "SET_TENANT_ID",
  },
]
`;

exports[`whoamiSagas get the user data from the API and save the tenantId 1`] = `
Array [
  Object {
    "data": Object {
      "cognitoId": "someCognitoId",
      "id": "someUserId",
      "tenantId": "someTenantId",
    },
    "type": "GET_WHOAMI_SUCCESS",
  },
  Object {
    "type": "INIT_WHOAMI_OBSERVER",
    "whoamiId": "someUserId",
  },
  Object {
    "data": "someTenantId",
    "type": "SET_TENANT_ID",
  },
]
`;
